### ISSUE TITLE
Determine testing standards

### ISSUE THREAD

**Issue by asclines** [TTUSDC/CPCEEDWebAppBackend#19]
- [Mocha & Chai](https://scotch.io/tutorials/test-a-node-restful-api-with-mocha-and-chai) to handle unit testing and asserts.
- [Mockgoose](https://github.com/mccormicka/Mockgoose) for handling mocking mongoose
- [Supertest](https://github.com/visionmedia/supertest) for handling testing the express server.
- [Mockery](https://www.npmjs.com/package/mockery) for mocking functions and modules

**Comment by asclines** [comment]
Closing as we have agreed on these testing frameworks.

**Comment by the-pat** [comment]
I move to reopen this issue.

@asclines @jmtaber129 @zachlefevre @ryanfaulkenberry100 

---

After looking at how unit tests are currently written within our code, we need set a few standards.

# Multiple asserts vs single asserts per unit test

I am _for_ single asserts. IMO, a unit test should only test a single condition. Ideally, a broken test should point to a single line of broken logic.

# Test naming conventions

I like the naming convention: `function_scenario_result`. This provides an verbose method of determining the error that occurred without having to look at the test.

# Test structure

I also like the following structure for unit tests:
```
/// Arrange
// Set up test here
...
/// Act
// Call the function getting tested
...
/// Assert
// Ensure scenario occurred as expected
...
```

Any thoughts?

**Comment by jmtaber129** [comment]
I agree with the test structure, but I disagree with the single asserts.  For methods that do multiple things (e.g. create an object in the DB and return the created object), the test should check each of those things.

**Comment by the-pat** [comment]
> For methods that do multiple things (e.g. create an object in the DB and return the created object), the test should check each of those things.

I think that should be broken out into two tests. One for creating an object and the other for getting the object from the DB.

**Comment by asclines** [comment]
In response to [this](https://github.com/TTUSDC/CPCEEDWebAppBackend/issues/19#issuecomment-292264697) comment.
### Multiple asserts vs single asserts per unit test
I disagree. A test can have multiple asserts no problem. @jmtaber129 makes a good example [above](https://github.com/TTUSDC/CPCEEDWebAppBackend/issues/19#issuecomment-292278114) for why.
To expand on that, I believe that a test should fail for one reason only, but that doesn't imply only one assert a test.


### Test naming conventions
I'm not quite sure what you mean. With Mocha's `describes()` and `it()` methods, one is more than able to fully detail what is being tested. What we put in there _does need to be standardized_, but @jmtaber129's tests in PR# 26 should be a good start at least.
For some good use cases of Mocha testing, checkout AirBnB's [react-date component testing](https://github.com/airbnb/react-dates/tree/master/test/components).
[WebSocket.io](https://github.com/LearnBoost/websocket.io/blob/master/test/server.js) provides an example of testing using multiple asserts.
As does [Express](https://github.com/expressjs/express/blob/master/test/app.render.js)


Yes, AirBnB's does only have one assert per test and I'm not saying thats a bad thing. I do see the argument for *striving* towards one assert per test, I do not believe we should *require* it. 


### Test structure 
I completely agree, [Arrange Act Asset](http://wiki.c2.com/?ArrangeActAssert) should be our goal in how we structure our tests.



In response to [this](https://github.com/TTUSDC/CPCEEDWebAppBackend/issues/19#issuecomment-292281015) comment, that does not always make sense. In the scenario @jmtaber129 describes, even though there are multiple assertions, it would be trivial to determine the point of failure. 



### For further reading
- [Is it OK to have multiple asserts in a single unit test?](http://softwareengineering.stackexchange.com/a/171502/153691)
- [Is it bad practice to have more than one assertion in a unit test?](http://stackoverflow.com/questions/762512/is-it-bad-practice-to-have-more-than-one-assertion-in-a-unit-test)
